### PR TITLE
Clean up transactions by making class

### PR DESCRIPTION
**Pull request by nhunzaker**
Signal felt pretty gnarly, so I cleaned it up a bit my moving some of the logic out into a `Transaction` class. I also added "states" to Transactions to make it clearer what is going on.

### CODE REVIEW

File: src/Microcosm.js [hunk 1]
```diff
@@ -114,26 +111,13 @@ Microcosm.prototype = {
...
 
-      transaction.done = transaction.done || done
-
-      this.rollforward()
+    this.transactions.push(transaction)
 
-      return body
-    }, action.apply(this, params))
+    return transaction.run(params)
   },
```
Comment (mzlock) [inline src/Microcosm.js]: such spaces between everything

Comment (nhunzaker) [inline src/Microcosm.js]: Such read, :book:

**Comment by mzlock** [comment]
:+1:

File: src/Microcosm.js (hunk 1 above)
Comment (nhunzaker) [inline src/Microcosm.js]: Actually this still feels a bit awkward. Transactions are event emitters, but they only have one event. I had considered just doing:

`new Transaction(action, params, this.rollforward.bind(this)`

Comment (nhunzaker) [inline src/Microcosm.js]: Maybe I should just do that...

Comment (nhunzaker) [inline src/Microcosm.js]: Well hmm. That's equally as awkward. I guess I'll leave it alone for now.

**Pull request merged**